// @dcloudio/uni-mp-compiler/dist/transforms/transformVBindAttrs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transformVBindAttrs = void 0;
const compiler_core_1 = require("@vue/compiler-core");
const uni_cli_shared_1 = require("@dcloudio/uni-cli-shared");
const transformVBindAttrs = (node, context) => {
    if (!(0, uni_cli_shared_1.isElementNode)(node)) {
        return;
    }
    const props = node.props;
    for (let i = 0; i < props.length; i++) {
        const prop = props[i];
        if (prop.type === compiler_core_1.NodeTypes.DIRECTIVE &&
            prop.name === 'bind' &&
            !prop.arg &&
            prop.modifiers.length === 0) {
            if (prop.exp &&
                prop.exp.type === compiler_core_1.NodeTypes.SIMPLE_EXPRESSION &&
                prop.exp.content === '$attrs') {
                const content = prop.exp.content;
                const newProps = [];
                // :class
                const classIndex = props.findIndex((p) => p.type === compiler_core_1.NodeTypes.DIRECTIVE &&
                    p.name === 'bind' &&
                    (0, compiler_core_1.isStaticArgOf)(p.arg, 'class'));
                const classProp = classIndex > -1 ? props[classIndex] : undefined;
                if (classProp && classProp.exp?.type === compiler_core_1.NodeTypes.SIMPLE_EXPRESSION) {
                    classProp.exp.content =
                        classIndex < i
                            ? `[${classProp.exp.content}, ${content}.class]`
                            : `[${content}.class, ${classProp.exp.content}]`;
                }
                else {
                    newProps.push(createBindDirective('class', `${content}.class`, prop.loc));
                }
                // :style
                const styleIndex = props.findIndex((p) => p.type === compiler_core_1.NodeTypes.DIRECTIVE &&
                    p.name === 'bind' &&
                    (0, compiler_core_1.isStaticArgOf)(p.arg, 'style'));
                const styleProp = styleIndex > -1 ? props[styleIndex] : undefined;
                if (styleProp && styleProp.exp?.type === compiler_core_1.NodeTypes.SIMPLE_EXPRESSION) {
                    styleProp.exp.content =
                        styleIndex < i
                            ? `[${styleProp.exp.content}, ${content}.style]`
                            : `[${content}.style, ${styleProp.exp.content}]`;
                }
                else {
                    newProps.push(createBindDirective('style', `${content}.style`, prop.loc));
                }
                // @click (only add if not already defined)
                const clickIndex = props.findIndex((p) => p.type === compiler_core_1.NodeTypes.DIRECTIVE &&
                    p.name === 'on' &&
                    (0, compiler_core_1.isStaticArgOf)(p.arg, 'click'));
                if (clickIndex > -1) {
                    const clickProp = props[clickIndex];
                    if (clickProp &&
                        clickProp.exp?.type === compiler_core_1.NodeTypes.SIMPLE_EXPRESSION) {
                        clickProp.exp.content =
                            clickIndex < i
                                ? `[${clickProp.exp.content}, ${content}.onClick]`
                                : `[${content}.onClick, ${clickProp.exp.content}]`;
                    }
                }
                else {
                    newProps.push(createOnDirective('click', `${content}.onClick`, prop.loc));
                }
                // :id
                // 查找后面是否还有 id，如果有，则忽略当前 id
                const hasId = props
                    .slice(i + 1)
                    .some((p) => (p.type === compiler_core_1.NodeTypes.ATTRIBUTE && p.name === 'id') ||
                    (p.type === compiler_core_1.NodeTypes.DIRECTIVE &&
                        p.name === 'bind' &&
                        (0, compiler_core_1.isStaticArgOf)(p.arg, 'id')));
                if (!hasId) {
                    newProps.push(createBindDirective('id', `${content}.id`, prop.loc));
                }
                props.splice(i, 1, ...newProps);
                i += newProps.length - 1;
            }
        }
    }
};
exports.transformVBindAttrs = transformVBindAttrs;
function createBindDirective(name, value, loc) {
    return {
        type: compiler_core_1.NodeTypes.DIRECTIVE,
        name: 'bind',
        modifiers: [],
        loc,
        arg: (0, compiler_core_1.createSimpleExpression)(name, true, loc),
        exp: (0, compiler_core_1.createSimpleExpression)(value, false, loc),
    };
}
function createOnDirective(name, value, loc) {
    return {
        type: compiler_core_1.NodeTypes.DIRECTIVE,
        name: 'on',
        modifiers: [],
        loc,
        arg: (0, compiler_core_1.createSimpleExpression)(name, true, loc),
        exp: (0, compiler_core_1.createSimpleExpression)(value, false, loc),
    };
}
